### PR TITLE
Feature/pricing

### LINKED ISSUE #214: Pricing updates
Transferred from https://github.com/openmsupply/remote-server/issues/947
A couple of points in https://github.com/openmsupply/org-issues/discussions/39

- [ ] Inbound line tax update
- [ ] inbound totals not updated
- [ ] inbound shipment pricing updates same as outbound shipment

After some discussion with @joshxg

Logic for pricing:

**Lines (all apart form unallocated line)**

- ~~#217 total_after_tax is always calculated (remove from mutation parameter), whenever the following is change~~
    - total_before_tax
    - tax
- #260 total_before_tax is still editable
    - for all lines apart from service line total_before_tax is calculated based on cost/sell price * number_of_packs
    - total_before_tax is still editable and will overwrite above calculation if it's specified in update/insert mutations (to be optional in insert mutations)
    - total_before_tax is recalculated (unless is specified in mutation) based on changes in:
        - number_of_packs
        - cost_price_per_pack (for inbound shipment)
        - stock_line_id (for outbound shipment)
        - item_id (for outbound shipment)
- tax is still editable

Result mutations (fields that affect pricing)

Outbound Line Insert
- total_before_tax (optional), overwrites calculation
- tax (optional)
- number_of_packs
- stock_line_id/item_id

Outbound Line Update
- total_before_tax (optional), overwrites calculation
- tax (optional)
- number_of_packs (optional)
- stock_line_id/item_id (optional)

Outbound Line Service Insert
- total_before_tax
- tax (optional)

Outbound Line Service Update
- total_before_tax (optional)
- tax (optional)

Inbound Line Insert
- total_before_tax (optional), overwrites calculation
- tax (optional)
- cost_price_per_pack
- number_of_packs

Inbound Line Update
- total_before_tax (optional), overwrites calculation
- tax (optional)
- cost_price_per_pack (optional)
- number_of_packs (optional)

Inbound Line Service Insert
- total_before_tax
- tax (optional)

Inbound Line Service Update
- total_before_tax (optional)
- tax (optional)

### PR DESCRIPTION
**Pull request by roxy-dao**
Closes #214

### CODE REVIEW

**Comment by andreievg** [comment]
@roxy-dao can you suggest some manual use cases for testers ? I think it's worth utilising our testing capacity for this one

**Comment by roxy-dao** [comment]
> @roxy-dao can you suggest some manual use cases for testers ? I think it's worth utilising our testing capacity for this one

- Creating new invoices with/without tax, and the total after tax should change accordingly. 
- Same when adding a service. 
- Update pack size of existing invoice to check for price change.

**Review by clemens-msupply — COMMENTED** [review]
Just that I don't forget over the weekend. Have some question regarding the calculations, best to do it next week in the office...

**Comment by andreievg** [comment]
@clemens-msupply 
> Just that I don't forget over the weekend. Have some question regarding the calculations, best to do it next week in the office...

Also good to check out this issue: https://github.com/openmsupply/remote-server/issues/947

File: server/service/src/invoice_line/outbound_shipment_line/update/generate.rs [hunk 1]
```diff
@@ -115,7 +114,7 @@ fn generate_line(
         item_code,
         stock_line_id: Some(stock_line_id),
         total_before_tax,
-        total_after_tax,
+        total_after_tax: total_after_tax(total_before_tax, tax),
```
Comment (clemens-msupply) [inline server/service/src/invoice_line/outbound_shipment_line/update/generate.rs:117]: total_after_tax should be calculated at the end after calculating total_before_tax, right?

Comment (roxy-dao) [reply server/service/src/invoice_line/outbound_shipment_line/update/generate.rs]: [17e41f8](https://github.com/openmsupply/open-msupply/pull/441/commits/17e41f8d21a1ae9ed9eb55f59dae07b9cf5f3840)

Comment (roxy-dao) [reply server/service/src/invoice_line/outbound_shipment_line/update/generate.rs]: [c56dbde](https://github.com/openmsupply/open-msupply/pull/441/commits/c56dbde13527207189fb6bc64515b8e3ae0bedc7)

**Review by clemens-msupply — APPROVED** [review]

**Pull request merged**